### ISSUE TITLE
Does NFS autofs work with auto_tmpdir?

### ISSUE THREAD

**Issue by OleHolmNielsen** [University-of-Delaware-IT-RCI/auto_tmpdir#6]
In [Bug 14344](https://bugs.schedmd.com/show_bug.cgi?id=14344) - job_container/tmpfs jobs fail because of race condition with NFS autofs mounting
we're discussing the issue that NFS automounted home directories do not work when using mount namespaces in the  job_container/tmpfs plugin.

Question: Have you tried out the auto_tmpdir plugin together with NFS automounts? I'm hoping that this solution would work for us.

**Comment by jtfrey** [comment]
Our user home directories are automounted NFS shares, and we have not seen any issues w.r.t. the auto_tmpdir plugin's hiding automount mountpoints:

```
[frey@login01.caviness ~]$ cd /lustre/scratch/frey/
[frey@login01.caviness frey]$ workgroup -g it_nss
[(it_nss:frey)@login01.caviness frey]$ sbatch --partition=devel --gres=gpu:1
#!/bin/bash

ls -latr ~

Submitted batch job 13861524

... job runs on node where ~ was NOT mounted prior to execution ...

[(it_nss:frey)@login01.caviness frey]$ head -10 slurm-13861524.out 
total 1839715
drwxr-----  3 frey everyone              3 May 25  2018 .pki
drwxr-xr-x  3 frey everyone              3 May 25  2018 .cmake
-rw-r--r--  1 frey everyone            595 Jun 26  2018 .travis.yml
drwx------  3 frey everyone              3 Jul  3  2018 .dbus
drwxr-xr-x  3 frey everyone              6 Jul  3  2018 .subversion
drwxr-xr-x  4 frey everyone              4 Jul  3  2018 .java
-rw-------  1 frey everyone             16 Jul  3  2018 .esd_auth
-rw-r--r--  1 frey everyone             44 Jul 11  2018 .valetrc
-rw-r--r--  1 frey everyone              0 Aug 20  2018 .moldenrc
```

The bind mount() calls in auto_tmpdir do differ from those used in the job_container/tmpfs plugin — the former remounts root in recursive+shared mode then clones a new namespace, while the latter clones a new namespace then remounts root recursive+private mode in it.  Seems like that order of operations might be to blame.

Having looked at the code now, one issue I can see with the job_container/tmpfs plugin is that it creates a ".ns" file under the per-job directory that's then used to communicate the namespace association to processes put into the "container" (via setns()).  If you've got a network-shared file system involved and a job on multiple nodes (e.g. MPI), each node's slurmd/slurmstepd will attempt to setup and use the same ".ns" file — e.g. `/home/niflheim/ohni/.ns` — for this purpose.  That's bound to lead to some problems.

**Comment by jtfrey** [comment]
#8 got me thinking:  if you have root_squash enabled on the NFS share and the job_container/tmpfs plugin in slurmd/slurmstepd (running as root) is attempting to create a new directory on that NFS share in a path not owned by root, it will fail to do so.

**Comment by OleHolmNielsen** [comment]
I have now built the ```auto_tmpdir.so``` shared library with this configuration:
```
cmake -DSLURM_PREFIX=/usr -DCMAKE_BUILD_TYPE=Release -DAUTO_TMPDIR_DEFAULT_LOCAL_PREFIX=/scratch/slurm- ..
```
and distributed the library using rsync to the ```/usr/lib64/slurm/``` directory on the compute nodes.
I've created ```/etc/slurm/plugstack.conf``` with the contents:
```
required    auto_tmpdir.so          mount=/tmp mount=/var/tmp
```

Now one must tell ```slurmd``` to pick up the changes on all compute nodes:
```
systemctl restart slurmd
```
I've submitted some batch jobs, and the plugin seems to work perfectly:
1. The /tmp, /var/tmp, and /dev/shm are bind mounted during the job and are removed when the job completes.
   The superuser can see the folder /scratch/slurm-<jobid>/ with the expected contents during the job.
2. The NFS automounted user home-directory is mounted correctly and is available within the job, in contrast to the Slurm [job_container.conf](https://slurm.schedmd.com/faq.html#tmpfs_jobcontainer) plugin.

So I think we're good with the auto_tmpdir plugin, thanks a lot!
